File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution:
    def isPalindrome(self, s, l, r, maxLen,i1,i2):
        n = len(s)
        while l>=0 and r<n and s[l] == s[r]:
            if (r-l+1) > maxLen:
                maxLen = r-l+1
                i1,i2 = l,r
            l -= 1
            r += 1
        return maxLen, i1, i2

    def longestPalindrome(self, s: str) -> str:
        ans = ""
        i1,i2 = 0,0
        maxLen = 0
        n = len(s)
        for i in range(n):
            #odd palindrome
            maxLen, i1,i2 = self.isPalindrome(s, i, i, maxLen, i1, i2)
            #even palindrome
            maxLen, i1,i2 = self.isPalindrome(s, i, i+1, maxLen, i1, i2)
        return s[i1:i2+1]
            
        

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
public:
     int isPalindrome(string s, int left, int right){

        while(left>=0 && right < s.size() && s[left]==s[right]){
            left--;
            right++;
        }
        return right-left-1;
    }


    string longestPalindrome(string s) {

        int n = s.size();
        int l = 0, r = 0;
        for(int i=0;i<n;i++){
            //odd palindrome
            int i1 = isPalindrome(s, i, i); 
            //even palindrome
            int i2 = isPalindrome(s, i, i+1);
            int maxLen = max(i1, i2);

            if(maxLen > (r-l)){
                l = i - (maxLen-1)/2;
                r = i + maxLen/2;
            }
        }
        string ans = "";
        for(int i=l;i<=r;i++){
            ans = ans+s[i];
        }
        return ans;
    }

};